README.md
# Peppermint Settings Panel

> Author: Mark Greaves <mark@peppermintos.com>


## License
> GPLv3, see debian/copyright.


debian/changelog
peppermint-settings-panel (0.0.4) trusty; urgency=low

  * Fix missing /usr/bin/peppermint-settings-panel script

 -- Mark Greaves (PCNetSpec) <mark@pcnetspec.co.uk>  Thu, 29 Oct 2015 13:25:00 +0000

peppermint-settings-panel (0.0.3) trusty; urgency=low

  * Add "install" file to copy peppermint-settings-panel script to /usr/bin

 -- Mark Greaves (PCNetSpec) <mark@pcnetspec.co.uk>  Thu, 29 Oct 2015 13:05:00 +0000

peppermint-settings-panel (0.0.3) trusty; urgency=low

  * Fix control file

 -- Mark Greaves (PCNetSpec) <mark@pcnetspec.co.uk>  Thu, 29 Oct 2015 12:27:00 +0000

peppermint-settings-panel (0.0.2) trusty; urgency=low

  * Fix Rules

 -- Mark Greaves (PCNetSpec) <mark@pcnetspec.co.uk>  Thu, 29 Oct 2015 12:05:00 +0000

peppermint-settings-panel (0.0.1) trusty; urgency=low

  * First build of peppermint-settings-panel

 -- Mark Greaves (PCNetSpec) <mark@pcnetspec.co.uk>  Thu, 29 Oct 2015 12:00:00 +0000


usr/share/peppermint-settings-panel/peppermint_sp.py
#!/usr/bin/env python

import gtk
import time
import os
import sys
from subprocess import Popen, PIPE
import shlex

from desktop_tool import DesktopToolWidget
from desktop_tool import get_icon as get_icon

from peppermint_sp_parse_xml import Peppermint_sp

#=======================================================================

class Config:
    # Position of NoteBook tabs gtk.POS_TOP or gtk.POS_LEFT
    TAB_POS = gtk.POS_LEFT
    # Number of columns
    NCOLUMNS = 3
    # Sizes
    ICON_SIZE = 48
    MAIN_WINDOW_BORDER_WIDTH = 5
    BTN_WIDTH = 150
    BTN_HEIGHT = 50
    # Icons
    MAIN_WINDOW_ICON = "preferences-desktop"
#=======================================================================

DEBUG = False
def debug(msg):
    if DEBUG:
        print "===> debug: " + msg

class Notebook(gtk.Notebook):

    def __init__(self):
        gtk.Notebook.__init__(self)
        self.set_tab_pos(Config.TAB_POS)

class Table(gtk.Table) :
    def __init__(self):
        gtk.Table.__init__(self)
        
    def attach(self, row, col, widget):
        # attach(child, left_attach, right_attach, top_attach, bottom_attach,..)
        gtk.Table.attach(self, widget, col, col + 1, row, row + 1)
    
class CategoryFrame(gtk.Frame) :

    def __init__(self, category):
        self.btn_height = Config.BTN_HEIGHT
        self.btn_width = Config.BTN_WIDTH
        
        gtk.Frame.__init__(self)

        self.alignment = gtk.Alignment(0.0, 0.0, 1.0, 1.0)
        self.alignment.set_padding(10, 10, 10, 10)
        self.add(self.alignment)
       
        self.table = Table()
        self.alignment.add(self.table)
        
        self.tool_button = []

        self.load_tools(category)

        
        
    def load_tools(self, category):
        
        row = 0
        col = 0
        i = 0
        for tool in category.tools:
            if Config.TAB_POS == gtk.POS_LEFT:
                orientation =  gtk.ORIENTATION_VERTICAL
                wrap_len = 20
            else:
                orientation =  gtk.ORIENTATION_HORIZONTAL
                wrap_len = 0 # no wrap
            icon_button = DesktopToolWidget(tool.label, tool.icon, Config.ICON_SIZE, orientation, wrap = wrap_len)
            self.tool_button.insert(i, icon_button)
            self.tool_button[i].set_size_request(self.btn_width, self.btn_height)
            self.tool_button[i].set_callback(self.on_item_activated, self, tool.action)

            self.table.attach(row, col, self.tool_button[i])
            self.table.set_col_spacing(col, 10)
            self.table.set_row_spacing(row, 10)
            #print "=== %s => i: %d row: %d col: %d" % (tool.label, i, row, col)
            i += 1
            col += 1
            if col == Config.NCOLUMNS:
                col = 0
                row += 1 
            
        #self.table.set_row_spacing(row, 50)
        
    def on_item_activated(self, action):
        print "on_item_activated: button [%s]" % action
        # Convert utf-8 string to ascii
        udata = action.decode("utf-8")
        asciidata = udata.encode("ascii","ignore")

        # Build the args list expected by Popen
        args = shlex.split(asciidata)
        # And start action in a subprocess
        print args
        process = Popen(args, stdout=PIPE, stderr=PIPE)


class MainWindow:
    
    def __init__(self):
        window = gtk.Window(gtk.WINDOW_TOPLEVEL)
        window.set_position(gtk.WIN_POS_CENTER)
        window.set_border_width(Config.MAIN_WINDOW_BORDER_WIDTH)
        pixbuf = get_icon(Config.MAIN_WINDOW_ICON, 16)
        window.set_icon(pixbuf)
        window.connect("delete_event", self.onDeleteEvent)

        # Create a new notebook
        self.notebook = Notebook()

        # Add the tab pages
        cat_index = 0
        
        peppermint_sp = Peppermint_sp("peppermint_sp.xml")
        #print peppermint_sp.to_string()

        window.set_title(peppermint_sp.title)

        for category in peppermint_sp.categories:
            
            category_name = category.title
            
            if True:
                tab_box = gtk.HBox(False, 2)
                tab_box.set_spacing(8)
                tab_label = gtk.Label(category_name)
                cat_index += 1

                pixbuf = get_icon(category.icon, Config.ICON_SIZE)
                tab_icon = gtk.Image()
                tab_icon.set_from_pixbuf(pixbuf)
                
                tab_icon.set_size_request(-1, 60)
                tab_box.pack_start(tab_icon, False)
                tab_box.pack_start(tab_label, False)

                # needed, otherwise even calling show_all on the notebook won't
                # make the hbox contents appear.
                tab_box.show_all()

            else:
                # tentative: ne convient pas: l'onglet a un ombrage de style bouton...
                tab_box = DesktopToolWidget(category_name, category.icon, Config.ICON_SIZE, gtk.ORIENTATION_HORIZONTAL)
            
            for tool in category.tools:
                frame = CategoryFrame(category)
                #assert frame.get_parent()
                #assert tab_box.get_parent()
                frame.show_all()
            
            self.notebook.append_page(frame, tab_box)
        
        # Set what page to start at
        self.notebook.set_current_page(0)
        self.notebook.show_all()
        window.add(self.notebook)
        
        window.show_all()

    def onDeleteEvent(self, widget, event=None, *arguments, **keywords):
        gtk.main_quit()
        return False


def main():
  gtk.main()
  return 0


if __name__ == "__main__":
    try:
        MainWindow()
        main()
    except KeyboardInterrupt:
        print "\nBye."
        sys.exit()


usr/share/peppermint-settings-panel/peppermint_sp_parse_xml.py
#!/usr/bin/python

from xml.dom.minidom import parse

filename = "peppermint_sp.xml"

class Peppermint_spTool:

    def __init__(self, label, icon, action):
        self.label = label
        self.icon = icon
        self.action = action
        pass

    def to_string(self):
        str = "    tool: label: %s icon: %s action: [%s]" % (self.label, self.icon, self.action)
        return str

class Peppermint_spCategory:

    def __init__(self, title, icon):
        self.title = title
        self.icon = icon
        self.tools = []
        pass

    def to_string(self):
        str = "category: title: %s icon: %s\n" % (self.title, self.icon)
        for tool in self.tools:
            str += tool.to_string() + '\n'
        return str

class Peppermint_sp:

    def __init__(self, xml_filename=""):
        self.categories = []
        if xml_filename:
            self.parse(xml_filename)
        
    def parse(self, xml_filename):
        self.dom = parse(xml_filename)
        root = self.dom.getElementsByTagName("peppermint_sp")[0]
        self.title = root.getAttribute('title')
        self.icon = root.getAttribute('icon')
        category_nodes = self.dom.getElementsByTagName("category")
        for n1 in category_nodes:
            attrs = n1.attributes
            title = n1.getAttribute('title')
            icon = n1.getAttribute('icon')
            # print title + " " + icon
            category = Peppermint_spCategory(title, icon)
            self.categories.append(category)
            tool_nodes = n1.getElementsByTagName("tool")
            for n2 in tool_nodes:
                label = n2.getAttribute('label')
                icon = n2.getAttribute('icon')
                action = n2.getAttribute('action')
                tool = Peppermint_spTool(label, icon, action)
                category.tools.append(tool)

    def to_string(self):
        str = ""
        for cat in self.categories:
            str += cat.to_string()
        return str
            
    def list_categories(self):
        pass
        

if __name__ == "__main__":
    peppermint_sp = Peppermint_sp("peppermint_sp.xml")
    print peppermint_sp.to_string()


usr/share/peppermint-settings-panel/desktop_tool.py
#!/usr/bin/python

import gtk
import os
import textwrap
import gobject


# For debug
def get_icon_info(icon_name, size=48):
    icon_theme = gtk.icon_theme_get_default()
    try:
        flags = 0
        pixbuf = icon_theme.load_icon(icon_name, size, flags)
        icon_info = icon_theme.lookup_icon(icon_name, size, flags)
        sz = icon_info.get_base_size()
        fname = icon_info.get_filename()
        print "get_icon_info: icon_name: " + icon_name + " Filename: " + fname + " base size: %d" % sz
    except gobject.GError:
        print "get_icon_info: icon_name: " + icon_name + " not present in theme "
        pass


# Return a pixbuf from icon name
def get_icon(name, size):
    """ Return a gtk.gdk.Pixbuf """
    found = False
    if name.startswith('/'):
        try:
            # name is an abslolute pathname to a file
            pixbuf = gtk.gdk.pixbuf_new_from_file(name)
            pixbuf = pixbuf.scale_simple(size, size, gtk.gdk.INTERP_BILINEAR)
            found = True
        except:
            # print "get_icon: Cannot get pixbuf from file: " + name
            name = os.basename(name).splitext()[0]

    if not found:
        theme = gtk.icon_theme_get_default()
        try:
            pixbuf = theme.load_icon(name, size, gtk.ICON_LOOKUP_USE_BUILTIN)
        except:
            try:
                filename = os.path.join('/usr/share/pixmaps', name)
                try:
                    pixbuf = gtk.gdk.pixbuf_new_from_file(filename)
                    found = True
                except:
                    # print "get_icon: [%s] not present in theme" % name
                    pixbuf = theme.load_icon('applications-system', size, gtk.ICON_LOOKUP_USE_BUILTIN)
            except:
                pixbuf = theme.load_icon(gtk.STOCK_PREFERENCES, 16, 0)
    # print "get_icon(name, size): icon name: %s pixbuf size: %dx%d" % (name, pixbuf.get_width(), pixbuf.get_height())
    if pixbuf.get_width() < 48 or pixbuf.get_height() < 48:
        pixbuf = pixbuf.scale_simple(size, size, gtk.gdk.INTERP_BILINEAR)
    if pixbuf.get_width() > 48 or pixbuf.get_height() > 48:
        pixbuf = pixbuf.scale_simple(size, size, gtk.gdk.INTERP_BILINEAR)
    return pixbuf


# My widget: icon and label wrapped in a EventBox
class DesktopToolWidget(gtk.EventBox):

    def __init__(self, label, icon_name, icon_size=48,
                 orientation=gtk.ORIENTATION_VERTICAL, border=4,
                 wrap=0):

        self.callback = None

        gtk.EventBox.__init__(self)
        self.set_border_width(0)

        if orientation == gtk.ORIENTATION_VERTICAL:
            # Label under Image
            box = gtk.VBox(False, border)
            box.set_border_width(int(border/2))
            box.set_spacing(border*2)
            box.show()
        else:
            # Label right side of Image
            box = gtk.HBox(False, border)
            box.set_border_width(int(border/2))
            box.set_spacing(border*2)
            box.show()

        btn_image = gtk.Image()
        # get_icon_info(icon_name, icon_size)
        pixbuf = get_icon(icon_name, icon_size)
        btn_image.set_from_pixbuf(pixbuf)

        btn_image.show()
        btn_image.set_size_request(icon_size, icon_size)
        box.pack_start(btn_image, False, False, 5)

        if label is not None:
            self.btn_label = gtk.Label()
            self.btn_label.set_use_markup(True)
            self.btn_label.set_line_wrap(True)
            if '&' in label:
                label = label.replace('&', '&amp;')
            label = label.replace("\\n", '\n')
            if wrap:
                label = textwrap.fill(label, wrap)
            self.btn_label.set_markup(label)

            if orientation == gtk.ORIENTATION_VERTICAL:
                self.btn_label.set_alignment(0.5, 0)
                self.btn_label.set_justify(gtk.JUSTIFY_CENTER)
                # align = gtk.Alignment(1, 0, 0.5, 0.5)
                # align.show()
                # align.add(self.btn_label)
                box.pack_start(self.btn_label, False, False)
            else:
                self.btn_label.set_alignment(0.2, 1)
                align = gtk.Alignment(1, 0, 0.5, 0.5)
                align.show()
                align.add(self.btn_label)
                box.pack_start(align, False, False)

        # Manage click
        self.connect("button-press-event", self.on_button_press)
        # Manage selection
        self.connect("enter-notify-event", self.on_enter_notify)
        self.connect("leave-notify-event", self.on_leave_notify)

        self.add(box)
        self.show_all()

    def on_button_press(self, w, event):
        print "on_button_press button 1 pressed: no callback"
        if event.type == gtk.gdk.BUTTON_PRESS and event.button == 1:
            print w
            if not w.callback:
                print "button 1 pressed: no callback"
            else:
                print "*** call me ***"
                w.callback(self.data)

    def on_enter_notify(self, w, event):
        # Change background of the EventBox
        self.selected_bg = self.style.bg[gtk.STATE_SELECTED]
        self.modify_bg(gtk.STATE_NORMAL, self.selected_bg)
        # Change foreground of the Label
        self.selected_fg = self.btn_label.style.fg[gtk.STATE_SELECTED]
        self.btn_label.modify_fg(gtk.STATE_NORMAL, self.selected_fg)

    def on_leave_notify(self, w, event):
        # Restore defaults
        self.modify_bg(gtk.STATE_NORMAL, None)
        self.btn_label.modify_fg(gtk.STATE_NORMAL, None)

    def set_callback(self, callback, w, data):
        # print "set_callback: data: " + data
        self.widget = w
        self.callback = callback
        self.data = data

# Test me
if __name__ == "__main__":

    class Test:
        Value = 5

        def do_something(self, args):
            print "do_something: args: " + args + " value: %d" % self.Value

    window = gtk.Window()
    window.set_title("DesktopToolWidget Test")
    window.set_border_width(10)

    vbox = gtk.VBox()
    vbox.set_spacing(8)

    icon_size = 48

    # Button 1
    icon_button1 = DesktopToolWidget("<b>Hello</b>\nIconButton", 'vlc', icon_size)
    icon_button1.set_size_request(150, 150)
    vbox.pack_start(icon_button1)
    # Attach a callback
    test = Test()
    icon_button1.set_callback(test.do_something, icon_button1, "Hello")

    # Button 2
    label = "<b>Hello</b>\nIconButton Horizontal"
    icon_button2 = DesktopToolWidget(label, 'synaptic', icon_size, gtk.ORIENTATION_HORIZONTAL)
    icon_button2.set_size_request(200, 100)
    vbox.pack_start(icon_button2)

    # Button 3
    label = "<b>Hello</b>This is a very long label a very long label a very long label"
    icon_button3 = DesktopToolWidget(label, 'synaptic', icon_size, gtk.ORIENTATION_HORIZONTAL, wrap=30)
    icon_button3.set_size_request(200, 100)
    vbox.pack_start(icon_button3)

    window.add(vbox)
    vbox.show()
    window.show()
    gtk.main()
